File: src/main/java/brain/Frontmatter.java
package brain;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Map;

/** Read YAML frontmatter from Markdown files. */
public final class Frontmatter {
  private static final String DELIM = "---";

  private Frontmatter() {}

  public record Doc(Map<String, Object> meta, String body) {}

  /** Returns the parsed frontmatter and the body. meta is empty when there is none. */
  public static Doc read(Path path) throws IOException {
    String text = Files.readString(path);
    String[] lines = text.split("\n", -1);
    if (lines.length == 0 || !lines[0].strip().equals(DELIM)) {
      return new Doc(Map.of(), text);
    }
    int closing = -1;
    for (int i = 1; i < lines.length; i++) {
      if (lines[i].strip().equals(DELIM)) {
        closing = i;
        break;
      }
    }
    if (closing == -1) {
      return new Doc(Map.of(), text);
    }
    String raw = String.join("\n", Arrays.copyOfRange(lines, 1, closing));
    Map<String, Object> meta = Yamls.load(raw);
    String body = String.join("\n", Arrays.copyOfRange(lines, closing + 1, lines.length));
    return new Doc(meta, body);
  }
}

File: brain_tools/frontmatter.js
// Read YAML frontmatter from Markdown files.

import { readFileSync } from "node:fs";
import yaml from "js-yaml";

const DELIM = "---";

// Return { meta, body }. meta is {} when there is no frontmatter.
export function read(path) {
  const text = readFileSync(path, "utf-8");
  const lines = text.split("\n");
  if (lines.length === 0 || lines[0].trim() !== DELIM) {
    return { meta: {}, body: text };
  }
  let closing = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === DELIM) {
      closing = i;
      break;
    }
  }
  if (closing === -1) {
    return { meta: {}, body: text };
  }
  const raw = lines.slice(1, closing).join("\n");
  const meta = yaml.load(raw) || {};
  const body = lines.slice(closing + 1).join("\n");
  return { meta, body };
}
